libs/libc/machine/Kconfig

```diff
@@ -137,6 +137,50 @@ config LIBC_ARCH_ELF_64BIT
 	default n
 	depends on LIBC_ARCH_ELF
 
+config LIBC_ARCH_ALLOW_KERNEL_ONLY
+	bool
+	default n
+
+config LIBC_ARCH_ALLOW_USER_ONLY
+	bool
+	default n
+
+if BUILD_PROTECTED || BUILD_KERNEL
+
+choice LIBC_ARCH_USAGE
+	prompt "libc architecture-specific usage"
+	default LIBC_ARCH_BOTH
+	depends on LIBC_ARCH_ALLOW_KERNEL_ONLY || LIBC_ARCH_ALLOW_USER_ONLY
+
+config LIBC_ARCH_BOTH
+	bool "Use architecture-specific on kernel and userspace"
+	depends on LIBC_ARCH_ALLOW_KERNEL_ONLY
+	depends on LIBC_ARCH_ALLOW_USER_ONLY
```

Comment:
Unused option, what's the intended use for it?

Reply:
Well, the considerations of this comment are the same of the next one's (https://github.com/apache/nuttx/pull/9138#pullrequestreview-1406556561), so I'll answer here, alright?

Well, first of all, I tried to avoid the need of fixing other architectures and make this change very isolated. If none of the `LIBC_ARCH_ALLOW_KERNEL_ONLY` and `LIBC_ARCH_ALLOW_USER_ONLY` options are selected, the build system will use the arch-specific for both the kernel and the userspace (as it works currently). These two options, however, if selected by the chip, state that the chip allows using arch-specific's libc on kernel/userspace only (and that's why I used `_ALLOW_xxx_ONLY`). However, it doesn't select it. This is done by the `LIBC_ARCH_USAGE` choice.

That being said, if only one of the `LIBC_ARCH_ALLOW_xxx_ONLY` is selected, the choice will be selected automatically to use the arch-specific implementation on kernel/userspace only. There might exist, however, a situation that - for a specific chip - both options are valid: the chip allows arch-specific to be used only in the kernel, only in the userspace, or both in the kernel and the userspace (same effect of the current implementation). So, if the chip selects both the `LIBC_ARCH_ALLOW_KERNEL_ONLY` and `LIBC_ARCH_ALLOW_USER_ONLY` options, it's up to the user/`defconfig` to select which combination (using the `LIBC_ARCH_USAGE` choice) will be used and `LIBC_ARCH_BOTH` is necessary to be able to not select one of the `LIBC_ARCH_xxx_ONLY` choices automatically. You can think as a non-restrictive option.

Reply:
> Well, first of all, I tried to avoid the need of fixing other architectures and make this change very isolated.

This fear of changing common code is compromising the quality of the proposed change by making it counterintuitive.
If the PR instead implemented the idea proposed in https://github.com/apache/nuttx/pull/9138#discussion_r1181087221, it would result in a much cleaner and more straightforward implementation:
> 1) the default value being y or 2) these options being selected by all the chips (currently, all the chips "allow" arch's libc in both the kernel and the userspace).

Reply:
> If the PR instead implemented the idea proposed in https://github.com/apache/nuttx/pull/9138#discussion_r1181087221, it would result in a much cleaner and more straightforward implementation:

I didn't get which proposed idea should be implemented. 1) or 2)?

Reply:
> > If the PR instead implemented the idea proposed in https://github.com/apache/nuttx/pull/9138#discussion_r1181087221, it would result in a much cleaner and more straightforward implementation:
> 
> 
> 
> I didn't get which proposed idea should be implemented. 1) or 2)?

Sorry, I meant the option 2.

Reply:
Another possibility is to invert the logic and instead of "allowing" the config could "prevent" the enabling on kernel/userspace.

Reply:
That could work! It avoids needing to update all the chips: something like `LIBC_ARCH_FORBID_xxx` would fit... I'll try to update the proposal on Tuesday (after the holidays)

Reply:
I've updated the MR with your suggestions. Could you please take a look at it?